Implement ravel

ejguan opened the pull request:
Related #38349
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46098 Implement ravel**

Doc:
![image](https://user-images.githubusercontent.com/68879799/95611323-ae5cf380-0a2f-11eb-9b8e-56bf79ce68af.png)

Differential Revision: [D24253213](https://our.internmc.facebook.com/intern/diff/D24253213)

ejguan committed:
Implement ravel

[ghstack-poisoned]

ejguan commented:
TODO:
Add tests.

ejguan committed:
Update on "Implement ravel"


Doc:
![image](https://user-images.githubusercontent.com/68879799/95611323-ae5cf380-0a2f-11eb-9b8e-56bf79ce68af.png)


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1624,6 +1624,11 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
 }
 
+Tensor ravel(const Tensor& self) {
+  // Same as flatten() with start_dim as 0 and end_dim as -1.
+  return native::flatten(self, 0, -1);
```
nit: `self.reshape(-1)` avoids some of the computation that goes on inside `flatten`.

Calling `native::flatten` instead of `at::flatten` inside of here means that we won't actually get autograd support. Using at::flatten (or reshape as pointed out above) fixes that problem. You can test this with something like:
```
x = torch.randn(3, 3, requires_grad=True)
out = x.ravel().sum()
out.backward()  # I think this throws an error with the current implementation
```

ejguan replied:
You are right. It doesn't really need to go into `flatten`.
But, t's interesting that it doesn't throw any error and gradient for the tensor is correct when I used `native::flatten`.

zou3519 replied:
Oh, I see what happened now. native::flatten calls at::reshape, and at::reshape is supported by the autograd.

zou3519 commented on test/test_torch.py:
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
         def test_gather(self):
             self._test_gather(self, lambda t: t)
 
+        def test_ravel(self):
```
You should add an entry to https://github.com/pytorch/pytorch/blob/59414b359d3a624c63032140e0e1fbc654aa6f5f/torch/testing/_internal/common_methods_invocations.py#L593 to auto-generate an autograd test. You'll be able to see the autograd test by running `pytest test/test_autograd.py -v -k "ravel"` locally (the -k "ravel" tells pytest to filter tests by name)

zou3519 commented on test/test_torch.py:
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
+                flat = src.ravel()
+                self.assertEqual(flat.shape, torch.Size([625]))
+                self.assertEqual(src.view(-1), flat.view(-1))
+
```
There are two cases for torch.ravel(tensor):
1) tensor is a contiguous tensor (e.g., tensor.is_contiguous() returns True). Then tensor.ravel() will return a view of the original tensor
2) tensor is a non-contiguous tensor. Then tensor.ravel() will return a copy of the original tensor.

We haven't tested (2) yet, that would be nice to add here.

zou3519 replied:
To check if a Tensor is a view of another tensor, one can check `tensor._base`. For example:

```
x = torch.randn(3, 3)
x_view = x.view(-1)
x_copy = x.clone()
assert x_view._base is x
assert x_copy._base is None # because x_copy isn't a view
```

zou3519 commented on test/test_torch.py:
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
+            for src in tensors:
+                flat = src.ravel()
+                self.assertEqual(flat.shape, torch.Size([625]))
+                self.assertEqual(src.view(-1), flat.view(-1))
```
self.assertEqual checks equality of values. It doesn't actually check if `flat` is a view of `src` or not, which we do want to check, because that is part of the semantics for ravel()

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -6490,6 +6490,25 @@ def merge_dicts(*dicts):
+           r"""
+ravel(input) -> Tensor
+
+Return a contiguous flattened tensor.
```
It's important to note here that  "A copy is made only if needed."

zou3519 reviewed (commented):
This generally looks good to me. We should add some more test cases:
- we should add an autograd test to make sure the operator works with autograd
- we should add some more cases to the test_torch test to check the behavior of ravel() on a non-contiguous Tensor

ejguan committed:
Update on "Implement ravel"


Doc:
![image](https://user-images.githubusercontent.com/68879799/95611323-ae5cf380-0a2f-11eb-9b8e-56bf79ce68af.png)


[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -590,6 +590,7 @@ def method_tests():
         ('view', (S,), (S,), '1d', (False,)),
         ('view', (), (dont_convert(()),), 'scalar_to_scalar', (False,)),
         ('view', (), (1,), 'scalar_to_1d', (False,)),
+        ('ravel', (S, S, S), NO_ARGS, '', (True,)),
```
I think the (True,) should be (False,): this boolean represents whether or not this operator is supported by JIT autodiff, which is not the same thing as autograd. Since we didn't make any changes to the JIT in this PR to support `ravel` in JIT autodiff, setting this as (True,) should cause some of the JIT tests to fail.

In general it is OK for an operator to not be supported by JIT autodiff (in fact, most operators aren't supported). JIT autodiff is a symbolic differentiation pass that is useful for the JIT when it tries to fuse things in the backwards graph.

ejguan replied:
I see. I though autodiff is similar to autograd. Haha, I may need to also check the JIT.

ejguan committed:
Update on "Implement ravel"


Doc:
![image](https://user-images.githubusercontent.com/68879799/95611323-ae5cf380-0a2f-11eb-9b8e-56bf79ce68af.png)


[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -1887,6 +1887,41 @@ def _test_gather(self, cast, test_bounds=True):
+                nc_src = src.t()
+                nc_flat = nc_src.ravel()
+                self.assertEqual(nc_flat.shape, torch.Size([25]))
+                self.assertEqual(nc_src.reshape(-1), nc_flat.reshape(-1))
```
nit: No need to reshape nc_flat again. This should be:
`self.assertEqual(nc_flat, nc-src.reshape(-1)`

zou3519 commented on test/test_torch.py:
```diff
@@ -1887,6 +1887,41 @@ def _test_gather(self, cast, test_bounds=True):
+            self.assertEqual(flat0.shape, flat1.shape)
+
+            # Test both float tensor and quantized tensor
+            tensors = [torch.randn(5, 5),
```
nit: We should probably test more shapes. The implementation code for ravel is simple enough that I can believe it works for everything, but there's no harm in testing more

zou3519 replied:
One edge case we should test is a tensor with a zero sized dimension. For example, the following should be true:
```
x = torch.randn([0, 2, 3]) # tensor of size [0, 2, 3]
y = x.ravel()
assert y.shape = ()
```

zou3519 reviewed (approved):
LGTM. Some minor nits on testing.

Also, we should try to test a tensor with a zero size in some dimension. I think the implementation should support it.

ejguan committed:
Update on "Implement ravel"


Doc:
![image](https://user-images.githubusercontent.com/68879799/95611323-ae5cf380-0a2f-11eb-9b8e-56bf79ce68af.png)


[ghstack-poisoned]
